File: src/bin/00003.go
package main

import (
	"fmt"
	"lib"
)

const N = 600851475143

func main() {
	t := lib.StartTimer()

	for x, n := 2, N; ; x++ {
		if x == n {
			fmt.Printf("Largest prime factor = %d\n", x)
			break
		}
		n = f(n, x)
	}

	lib.StopTimer(t)
}

func f(n, x int) int {
	switch {
    case n == 1 || x == 1:
        return n
	case n%x == 0:
		return f(n/x, x)
    default :
        return n
	}
}

File: src/bin/00003.js
console.time("Largest prime factor")

function f(n, x) {
    if (n == 1 || x == 1)
        return 1
    else if (n % x == 0)
        return f(n / x, x)
    else
        return n
}

let n = 600851475143
let x

for (x = 2; x <= n; x++) {
    n = f(n, x)
    if (n == 1) break;
}

console.timeEnd("Largest prime factor")

console.log("Largest prime factor = %d", x)

console.assert(6857 == x, "Wrong!")